python/rascaline/rascaline/utils/clebsch_gordan/correlate_density.py

```diff
@@ -76,185 +92,332 @@ def correlate_density(
+                "Must be greater equal 0."
+            )
+        self._max_angular = max_angular
+
+        if self._cg_backend == "python-dense":
+            sparse = False
+            use_mops = False
+        elif self._cg_backend == "python-sparse":
+            sparse = True
+            use_mops = False
+        elif self._cg_backend == "mops":
+            sparse = True
+            use_mops = True
+
+        self._cg_coeffs = _cg_cache.ClebschGordanReal(
+            self._max_angular,
+            sparse=sparse,
+            use_mops=use_mops,
+            use_torch=(self._arrays_backend == "torch"),
+        ).coeffs
+
+        # Check inputs
+        if correlation_order <= 1:
+            raise ValueError("`correlation_order` must be > 1")
+        self._correlation_order = correlation_order
+
+        n_iterations = correlation_order - 1  # num iterations
+        # Parse the selected keys
+        self._angular_cutoff = angular_cutoff
+
+        if self._arrays_backend == "torch":
+            array_like = torch.empty(0)
+        elif self._arrays_backend == "numpy":
+            array_like = np.empty(0)
+
+        self._selected_keys: List[Union[Labels, None]] = (
+            _clebsch_gordan._parse_selected_keys(
+                n_iterations=n_iterations,
+                array_like=array_like,
+                angular_cutoff=self._angular_cutoff,
+                selected_keys=selected_keys,
+            )
         )
-    # Check metadata
-    if not (
-        _dispatch.all(density.keys.names == ["spherical_harmonics_l", "species_center"])
-        or _dispatch.all(
-            density.keys.names
-            == ["spherical_harmonics_l", "species_center", "species_neighbor"]
+        # Parse the bool flags that control skipping of redundant CG combinations
+        # and TensorMap output from each iteration
+        self._skip_redundant, self._output_selection = (
+            _clebsch_gordan._parse_bool_iteration_filters(
+                n_iterations,
+                skip_redundant=skip_redundant,
+                output_selection=output_selection,
+            )
         )
-    ):
-        raise ValueError(
-            "input `density` must have key names"
-            ' ["spherical_harmonics_l", "species_center"] or'
-            ' ["spherical_harmonics_l", "species_center", "species_neighbor"]'
+
+    @property
+    def correlation_order(self):
+        return self._correlation_order
+
+    @property
+    def selected_keys(self) -> List[Union[Labels, None]]:
+        return self._selected_keys
+
+    @property
+    def skip_redundant(self) -> List[bool]:
+        return self._skip_redundant
+
+    @property
+    def output_selection(self) -> List[bool]:
+        return self._output_selection
+
+    @property
+    def arrays_backend(self):
+        return self._arrays_backend
+
+    @property
+    def cg_backend(self):
+        return self._cg_backend
+
+    @property
+    def cg_coeffs(self) -> TensorMap:
```

Comment:
Do we really need to expose these as properties?

At the very least, `cg_coeffs` is an internal implementation detail